### ISSUE TITLE
Remove unwanted html elements with regex or xpaths

### ISSUE THREAD

**Issue by andremacola** [adbar/trafilatura#283]
Possibility to remove unnecessary html elements before starting the extraction process.

There are often some elements within the extracted text that are not article content.

Titles should by default not come inside the extracted text, or there should be an option to remove them (maybe this requires another issue)

Something like:

```python
unwanted = [
  'iframe',
  'button',
  'figcaption',
  'caption',
  'form',
  'aside',
  'script',
  'style',
  'ins',
  'link',
  'header',
  'footer',
  '#comments',
  'nav',
  '.post-comments',
  '.post-tags',
  '.wp-block-embed',
  '.wp-caption-text',
  'svg',
  '[class^=ads]',
  '[class*=ads-]',
  '[style="display:none"]',
  '[style*="display:none"]',
  '[style*="display: none"]',
  '[itemprop*="description"]',
  '.push-web-notification',
  '.mc-column.entities',
  '.newsletter-component',
  '.post-subject',
  '.post-info',
  '.addthis_tool',
  '.pt-cv-wrapper'
]

article = trafilatura.bare_extraction(document,
        unwanted_elements=unwanted
        include_comments=False, include_tables=False,
        favor_precision=True, favor_recall=True,
        no_fallback=True, target_language=None,
        date_extraction_params={'extensive_search': True, 'original_date': True, 'outputformat': "%Y-%m-%dT%H:%M:%S%z"},
        config=config)
```

**Comment by andremacola** [comment]
Here an option to achieve this

```python
import lxml.html
import trafilatura

unwanted_elements = [
  'iframe',
  'button',
  'figcaption',
  'caption',
  'form',
  'aside',
  'script',
  'style',
  'ins',
  'link',
  'header',
  'footer',
  '#comments',
  'nav',
  '.post-comments',
  '.post-tags',
  '.wp-block-embed',
  '.wp-caption-text',
  'svg',
  '[class^=ads]',
  '[class*=ads-]',
  '[style="display:none"]',
  '[style*="display:none"]',
  '[style*="display: none"]',
  '[itemprop*="description"]',
  '.push-web-notification',
  '.mc-column.entities',
  '.newsletter-component',
  '.post-subject',
  '.post-info',
  '.addthis_tool',
  '.pt-cv-wrapper'
]

def remove_unwanted_elements(html: str):
    document = lxml.html.document_fromstring(html)
    targets = ','.join(unwanted_elements)
    elements = document.cssselect(targets)

    for element in elements:
        element.drop_tree()

    return lxml.etree.tostring(document, encoding='unicode')

html = trafilatura.fetch_url(self.url, config=config)
document = remove_unwanted_elements(html)

article = trafilatura.bare_extraction(document,
        unwanted_elements=unwanted
        include_comments=False, include_tables=False,
        favor_precision=True, favor_recall=True,
        no_fallback=True, target_language=None,
        date_extraction_params={'extensive_search': True, 'original_date': True, 'outputformat': "%Y-%m-%dT%H:%M:%S%z"},
        config=config)
```

**Comment by adbar** [comment]
Hi @andremacola, it looks like a good idea but I have to test it thoroughly to make sure it works as expected.

The question whether titles should be included in the result is debatable but I understand.

The list you provide could also be shortened as certain expressions are already accounted for, notably here:
https://github.com/adbar/trafilatura/blob/34bb2ef9d45736ddad03bc3a26dd585dd169ef6d/trafilatura/htmlprocessing.py#L47
and here:
https://github.com/adbar/trafilatura/blob/34bb2ef9d45736ddad03bc3a26dd585dd169ef6d/trafilatura/xpaths.py#L99

**Comment by andremacola** [comment]
Oh, I didn't notice the `OVERALL_DISCARD_XPATH`. That's nice.

To avoid modifying core files, I think some constants can be included in settings.cfg of use_config() options

**Comment by adbar** [comment]
Yes, it makes sense to allow users to modify the list of elements to delete as part of the settings.

**Comment by andremacola** [comment]
> Here an option to achieve this
> 
> ```python
> import lxml.html
> import trafilatura
> 
> unwanted_elements = [
>   'iframe',
>   'button',
>   'figcaption',
>   'caption',
>   'form',
>   'aside',
>   'script',
>   'style',
>   'ins',
>   'link',
>   'header',
>   'footer',
>   '#comments',
>   'nav',
>   '.post-comments',
>   '.post-tags',
>   '.wp-block-embed',
>   '.wp-caption-text',
>   'svg',
>   '[class^=ads]',
>   '[class*=ads-]',
>   '[style="display:none"]',
>   '[style*="display:none"]',
>   '[style*="display: none"]',
>   '[itemprop*="description"]',
>   '.push-web-notification',
>   '.mc-column.entities',
>   '.newsletter-component',
>   '.post-subject',
>   '.post-info',
>   '.addthis_tool',
>   '.pt-cv-wrapper'
> ]
> 
> def remove_unwanted_elements(html: str):
>     document = lxml.html.document_fromstring(html)
>     targets = ','.join(unwanted_elements)
>     elements = document.cssselect(targets)
> 
>     for element in elements:
>         element.drop_tree()
> 
>     return lxml.etree.tostring(document, encoding='unicode')
> 
> html = trafilatura.fetch_url(self.url, config=config)
> document = remove_unwanted_elements(html)
> 
> article = trafilatura.bare_extraction(document,
>         unwanted_elements=unwanted
>         include_comments=False, include_tables=False,
>         favor_precision=True, favor_recall=True,
>         no_fallback=True, target_language=None,
>         date_extraction_params={'extensive_search': True, 'original_date': True, 'outputformat': "%Y-%m-%dT%H:%M:%S%z"},
>         config=config)
> ```

There is a problem doing this, maybe because of this: https://lxml.de/compatibility.html

With some htmls, lxml loose the reference of the element inside the tree and when you remove an unnecessary element, for example: a `<nav>` tag. There may be situations where the lxml wraps all the content inside the `<nav>` and you end up losing the html markup to extract the content.

This is not common, but happens. Here an example: https://macola.com.br/lxml-reference.html

If you print the document tree you see something like this:

![Captura de Tela 2023-02-27 às 13 02 31](https://user-images.githubusercontent.com/3408809/221615235-9ab76616-93b1-45ef-a8c0-fdb9d24c70be.jpg)

I'll be happy to do a patch if someone points some right direction (if this is not a lxml limitation)

Any clue @adbar ?

**Comment by andremacola** [comment]
Ok.. just realized that the example is missing some close </div> tags inside one of the `<nav>` tag

**Comment by adbar** [comment]
As you say the `nav` problem is more a LXML issue, I'm not sure how to fix it.

The `unwanted_elements` list could be a part of the extraction options. It would be loaded from the settings by default but could be supplied by the users.

**Comment by adbar** [comment]
Hi @andremacola, the extraction functions now accept a `prune_xpath` argument (XPath string or list) so I'm closing this issue.
